observatory_platform/telescopes/terraform.py

```diff
@@ -0,0 +1,144 @@
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# Author: Aniek Roelofs
+import logging
+from airflow.exceptions import AirflowException
+from airflow.hooks.base_hook import BaseHook
+from airflow.models import Variable
+from airflow.models.taskinstance import TaskInstance
+from airflow.models.dagrun import DagRun
+from observatory_platform.utils.terraform_utils import TerraformApi
+from observatory_platform.utils.config_utils import (check_variables, check_connections, AirflowVar, AirflowConn)
+
+
+class TerraformTelescope:
+    # DAG variables
+    DAG_ID_CREATE = "vm_create"
+    DAG_ID_DESTROY = "vm_destroy"
+
+    XCOM_WORKSPACE_ID = "workspace_id"
+    XCOM_START_TIME_VM = "start_time_vm"
+
+    TASK_ID_CHECK_DEPENDENCIES = "check_dependencies"
+    TASK_ID_WORKSPACE = "get_workspace_id"
+    TASK_ID_VAR_CREATE = "update_var_create"
+    TASK_ID_VAR_DESTROY = "update_var_destroy"
+    TASK_ID_RUN = "terraform_run_configuration"
+    TASK_ID_DAG_SUCCESS = "check_all_dags_success"
+    TASK_ID_STOP = "stop_workflow"
+
+    @staticmethod
+    def check_dependencies(**kwargs):
+        """ Check that all variables exist that are required to run the DAG.
+
+        :param kwargs: the context passed from the PythonOperator. See https://airflow.apache.org/docs/stable/macros-ref.html
+        for a list of the keyword arguments that are passed to this argument.
+        :return: None.
+        """
+
+        vars_valid = check_variables(AirflowVar.project_id.get(), AirflowVar.terraform_organization.get(),
+                                     AirflowVar.terraform_prefix.get())
+        conns_valid = check_connections(AirflowConn.terraform.get())
+
+        if not vars_valid or not conns_valid:
+            raise AirflowException('Required variables or connections are missing')
+
+    @staticmethod
+    def get_workspace_id(**kwargs):
+        token = BaseHook.get_connection(AirflowConn.terraform.get()).password
+        terraform_api = TerraformApi(token)
+
+        # Get organization
+        organization = Variable.get(AirflowVar.terraform_organization.get())
+
+        # Get workspace
+        prefix = Variable.get(AirflowVar.terraform_prefix.get())
+        environment = Variable.get(AirflowVar.environment.get())
+        workspace = prefix + environment
+
+        # Get workspace ID
+        workspace_id = terraform_api.workspace_id(organization, workspace)
+
+        # Push messages
+        ti: TaskInstance = kwargs['ti']
+        ti.xcom_push(TerraformTelescope.XCOM_WORKSPACE_ID, workspace_id)
```

Comment:
I wonder if it would be cleaner to wrap this functionality up as a function and get the workspace id, rather than using xcoms to get the workspace id within each task?

Reply:
Thanks for the idea, my reasoning was that I would try to limit the number of requests made to the terraform API, but you're right that it would be cleaner!

Reply:
See https://github.com/The-Academic-Observatory/observatory-platform/pull/231/commits/df09eec07da0812782fd780a4c45e5a0c0a976a3